backend/deep/subscription.py

```diff
@@ -0,0 +1,205 @@
+            # he has been added to.
+            redis_key = '{}-groups'.format(self.message.reply_channel)
+
+            # Subscribe or unsubscribe ?
+            action = content['action']
+
+            if not action:
+                raise Exception('Action not provided')
+
+            if action == 'subscribe':
+                # Subscribe to a group
+
+                # Step 1: validate, and encode the request
+                self.validate(content)
+                code = self.encode(content)
+
+                # Step 2: Add to subscription channel group
+                # and also store the group reference in the
+                # redis list for the client.
+                Group(code).add(self.message.reply_channel)
+                r.lpush(redis_key, code)
+
+                # Step 3: send a success reply along with group code
+                self.send({'code': code, 'success': True})
+
+            elif action == 'unsubscribe':
+                if content['channel'] == 'all':
+                    # Unsubscribe from all groups
+
+                    # For this, we use the redis group list for the client.
+                    # Iteratively, pop each group from the list,
+                    # remove the client from the group itself.
+
+                    # Also maintain list of codes of unsubscribed events.
+                    unsubscribed_codes = []
+
+                    code = r.lpop(redis_key)
+                    while code:
+                        code = code.decode('utf-8')
+
+                        Group(code).discard(self.message.reply_channel)
+                        unsubscribed_codes.append(code)
+
+                        code = r.lpop(redis_key)
+
+                    # If the list is empty, we may as well remove the list
+                    # from the redis store.
+                    if r.llen(redis_key) == 0:
+                        r.delete(redis_key)
+
+                    # Reply back with the unscubscribed_codes
+                    self.send({'unsubscribed_codes': unsubscribed_codes,
+                               'success': True})
+
+                else:
+                    # Unsubscribe from a single group
+                    # The request is in similar form as subscription,
+                    # so we get the code from the request.
+                    self.validate(content)
+                    code = self.encode(content)
+
+                    # Remove the client from the group
+                    # Then remove the group from the redis list for the client
+                    Group(code).discard(self.message.reply_channel)
+                    r.lrem(redis_key, 0, code)
+
+                    # If empty, remove the redis list
+                    if r.llen(redis_key) == 0:
+                        r.delete(redis_key)
+
+                    # Reply back with the unsubscribed_codes.
+                    self.send({'ubsubscribed_codes': [code], 'success': True})
```

Comment:
Is "ubsubscribed_codes" intentional or a typo?

Reply:
typo :laughing: